lib/puppet_x/templates/invoke_ps_command.erb

```diff
@@ -0,0 +1,284 @@
+$hostSource = @"
+using System;
+using System.Globalization;
+using System.Management.Automation.Host;
+using System.Threading;
+
+public class PuppetPSHost : PSHost
+{
+    private Guid _hostId = Guid.NewGuid();
+    private bool shouldExit;
+    private int exitCode;
+
+    public PuppetPSHost () {}
+
+    public bool ShouldExit
+    {
+        get { return this.shouldExit; }
+    }
+
+    public int ExitCode
+    {
+        get { return this.exitCode; }
+    }
+
+    public void ResetExitStatus()
+    {
+      this.exitCode = 0;
+      this.shouldExit = false;
+    }
+
+    public override Guid InstanceId
+    {
+        get { return _hostId; }
+    }
+
+    public override string Name
+    {
+        get { return "PuppetPSHost"; }
+    }
+
+    public override Version Version
+    {
+        get { return new Version(1, 0); }
+    }
+
+    public override PSHostUserInterface UI
+    {
+        get { return null; }
+    }
+
+    public override CultureInfo CurrentCulture
+    {
+        get { return Thread.CurrentThread.CurrentCulture; }
+    }
+
+    public override CultureInfo CurrentUICulture
+    {
+        get { return Thread.CurrentThread.CurrentUICulture; }
+    }
+
+    public override void EnterNestedPrompt()
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void ExitNestedPrompt()
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void NotifyBeginApplication()
+    {
+        return;
+    }
+
+    public override void NotifyEndApplication()
+    {
+        return;
+    }
+
+    public override void SetShouldExit(int exitCode)
+    {
+      this.shouldExit = true;
+      this.exitCode = exitCode;
+    }
+}
+"@
+
+function New-XmlResult
+{
+  param(
+    [Parameter()]$exitcode,
+    [Parameter()]$output,
+    [Parameter()]$errormessage
+  )
+
+  # we make our own xml because ConvertTo-Xml makes hard to parse xml ruby side
+  # and we need to be sure
+  [System.Xml.XmlDocument]$xDoc = New-Object System.Xml.XmlDocument
```

Comment:
So I think I understand the reasoning behind writing the stream in a specific manner back to Ruby (except for the indent), but I can't say that I understand building the structure like this... but maybe there was a reason? (I can't figure why all the casting though...)

This should work just fine and achieve the same result, if we really need to leave out children that are null:

```powershell
$xml = [xml]@"
<ReturnResult>
  <Property Name='exitCode'>$exitCode</Property>
</ReturnResult>
"@

if ($errormessage -ne $null)
{
  $prop = $xml.ReturnResult.FirstChild.Clone()
  $prop.Name = 'errormessage'
  $prop.InnerText = [Convert]::ToBase64String([Text.Encoding]::UTF8.GetBytes($errormessage))
  $xml.ReturnResult.AppendChild($prop) | Out-Null
}

if ($output -ne $null)
{
  $prop = $xml.ReturnResult.FirstChild.Clone()
  $prop.Name = 'stdout'
  $prop.InnerText = [Convert]::ToBase64String([Text.Encoding]::UTF8.GetBytes($output))
  $xml.ReturnResult.AppendChild($prop) | Out-Null
}
```

Even that feels a lot clunkier than necessary, and could really be something as simple as this, which is a lot easier to understand IMHO:

```powershell
$xml = [xml]@"
<ReturnResult>
  <Property Name='exitCode'>$exitCode</Property>
  <Property Name='errormessage'>$([Convert]::ToBase64String([Text.Encoding]::UTF8.GetBytes([string]$errormessage)))</Property>
  <Property Name='stdout'>$([Convert]::ToBase64String([Text.Encoding]::UTF8.GetBytes([string]$output)))</Property>
</ReturnResult>
"@
```

We can trivially change the Ruby receiving side here to deal with an empty string instead of a missing node for `errormessage` and `stdout`